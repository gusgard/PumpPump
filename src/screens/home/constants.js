//  name of this module
export const NAME = 'home'

//  action types
export const FETCH_USER = `${NAME}/FETCH_USER`
export const FETCH_PHOTOS_SLIDER = `${NAME}/FETCH_PHOTOS_SLIDER`
export const FETCH_PHOTOS_GRID = `${NAME}/FETCH_PHOTOS_GRID`
