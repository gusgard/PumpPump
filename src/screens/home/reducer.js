import { handleActions } from 'redux-actions'
import update from 'immutability-helper'

import {
  FETCH_USER,
  FETCH_PHOTOS_SLIDER,
  FETCH_PHOTOS_GRID,
} from './constants'

export const initialState = {
  user: {
    photos: [],
    bio: [],
    name: '',
    thumbnail: {},
  },
  popularPhotos: [],
}

export default handleActions(
  {
    [FETCH_USER]: (state, { payload: { user } }) =>
      update(state, {
        user: { $set: user },
      }),
    [FETCH_PHOTOS_GRID]: (state, { payload: { popularPhotos } }) =>
      update(state, {
        popularPhotos: { $set: popularPhotos },
      }),
    [FETCH_PHOTOS_SLIDER]: (state, { payload: { photos } }) =>
      update(state, {
        user: {
          photos: { $set: photos },
        },
      }),
  },
  initialState,
)
