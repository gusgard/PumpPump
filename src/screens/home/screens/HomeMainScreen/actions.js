import { Alert } from 'react-native';

import { UserFacade } from '@entities';

import {
  FETCH_USER,
  FETCH_PHOTOS_SLIDER,
  FETCH_PHOTOS_GRID,
} from '../../constants';

export const fetchUser = () => async dispatch => {
  try {
    const user = await UserFacade.fetchOne();
    dispatch({ type: FETCH_USER, payload: { user } });

    const photos = await UserFacade.fetchPhotos(user.id);
    dispatch({ type: FETCH_PHOTOS_SLIDER, payload: { photos } });
  } catch (e) {
    const title = 'Server error';
    const message = 'Fail connection to server';
    Alert.alert(title, message, [{ text: 'OK' }]);
  }
};

export const fetchPhotosGrid = grid => ({
  type: FETCH_PHOTOS_GRID,
  payload: { grid },
});
