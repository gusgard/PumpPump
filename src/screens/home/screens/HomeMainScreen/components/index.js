export { default as UserBio } from './UserBio';
