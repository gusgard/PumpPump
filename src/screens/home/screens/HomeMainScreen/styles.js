import { StyleSheet } from 'react-native';

import { width, colors } from '@themes';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.primary,
  },
  imageGrid: {
    width: '100%',
    height: '100%',
  },
  imageSwiper: {
    height: width,
    width,
  },
});
