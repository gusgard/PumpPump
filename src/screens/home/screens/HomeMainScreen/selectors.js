import { createSelector } from 'reselect';

import { NAME } from '../../constants';

const getPopularPost = state => state[NAME].popular;
export const getPopularPostState = createSelector(
  [getPopularPost],
  popular => popular,
);

const getUser = state => state[NAME].user;
export const getUserState = createSelector([getUser], user => user);
