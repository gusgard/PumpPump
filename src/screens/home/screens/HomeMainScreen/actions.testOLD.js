import mockStore from 'redux-mock-store'
import MockAdapter from 'axios-mock-adapter'

import { HttpService } from '@common'
import { Feed, User, FeedFacade, UserFacade } from '@entities'

import { fetchPhotosGrid, fetchUser } from './actions'
import { FETCH_USER, FETCH_PHOTOS_SLIDER, FETCH_ALL_POPULAR } from '../../constants'

const mockApi = new MockAdapter(HttpService.instance)

describe('home actions', () => {
  let store
  let posts
  let rawUser
  beforeEach(() => {
    store = mockStore()
    posts = [
      {
        createdAt: '2017-11-04T19:49:54.136Z',
        thumbnail: 'https://reactjs.org/logo-og.png',
        objectId: 18561785
      },
      {
        createdAt: '2017-11-04T19:31:41.206Z',
        thumbnail: 'https://reactjs.org/logo-og.png',
        objectId: 18561778
      }
    ]
    rawUser = {
      bio:
        'Motivation to become the best version of you!  💙💪🌎\n\nIt\'s #HealthyHolidays ❄️🏋️\u200d♀️\n\n' +
        'Share your photos all month long to be featured!\n\n👻 Snapchat @PumpUp\n\nGet your #TeamPumpUp gear ⬇️',
      name: 'pumpup',
      profileThumbnail: 'https://reactjs.org/logo-og.png',
      createdAt: '2014-02-03T07:21:44.372Z',
      objectId: 318381
    }
  })

  it('fetch popular photos', async () => {
    const popularPhotos = posts.map(item => User.decodePhoto(item))

    const expectedActions = [{ type: FETCH_ALL_POPULAR, payload: { popularPhotos } }]

    const response = { result: { posts } }

    mockApi.onPost(FeedFacade.fetchPopularPhotosUrl).reply(200, response)

    await store.dispatch(fetchPhotosGrid())

    const storeActions = store.getActions()
    expect(storeActions).toEqual(expectedActions)
    expect(storeActions).toMatchSnapshot()
  })

  it('fetch user data', async () => {
    const user = User.decode(rawUser)
    const photos = posts.map(item => Feed.decode(item))

    const expectedActions = [
      { type: FETCH_USER, payload: { user } },
      { type: FETCH_PHOTOS_SLIDER, payload: { photos } }
    ]

    const userId = 1
    mockApi.onPost(`${UserFacade.fetchOneUrl}${userId}`).reply(200, rawUser)

    const responseFetchPhotos = { result: { posts } }
    mockApi.onPost(UserFacade.fetchPhotosUrl).reply(200, responseFetchPhotos)

    await store.dispatch(fetchUser(userId))

    const storeActions = store.getActions()
    expect(storeActions).toEqual(expectedActions)
    expect(storeActions).toMatchSnapshot()
  })
})
