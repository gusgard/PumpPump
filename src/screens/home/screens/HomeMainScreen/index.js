import React, { PureComponent } from 'react';
import { ScrollView, View } from 'react-native';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Slider, ListGrid } from '@components';

import * as actions from './actions';
import styles from './styles';
import { getUserState, getPopularPhotosState } from './selectors';
import { UserProfile } from './components';

class HomeMainScreen extends PureComponent {
  static propTypes = {
    user: PropTypes.object.isRequired,
    popularPhotos: PropTypes.array.isRequired,
    fetchUser: PropTypes.func.isRequired,
    fetchPhotosGrid: PropTypes.func.isRequired,
  };

  componentWillMount() {
    const { fetchUser, fetchPhotosGrid } = this.props;
    const userId = 1234;
    fetchUser(userId);
    fetchPhotosGrid();
  }

  render() {
    const { user, popularPhotos } = this.props;
    return (
      <View style={styles.container}>
        <UserProfile
          name={user.name}
          description={user.description}
          picture={user.thumbnail}
        />
        <ScrollView>
          <Slider items={user.photos} />
          <ListGrid items={popularPhotos} />
        </ScrollView>
      </View>
    );
  }
}

export default connect(
  state => ({
    user: getUserState(state),
    popularPhotos: getPopularPhotosState(state),
  }),
  dispatch => ({
    fetchUser: id => dispatch(actions.fetchUser(id)),
    fetchPhotosGrid: () => dispatch(actions.fetchPhotosGrid()),
  }),
)(HomeMainScreen);
