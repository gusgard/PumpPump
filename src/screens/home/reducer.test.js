import home, { initialState } from './reducer';
import * as actions from './screens/HomeMainScreen/actions';

it('fetch popular photos', () => {
  expect(home(initialState, actions.fetchPhotosGrid())).toMatchSnapshot();
});

it('fetch user', () => {
  expect(home(initialState, actions.fetchUser())).toMatchSnapshot();
});
