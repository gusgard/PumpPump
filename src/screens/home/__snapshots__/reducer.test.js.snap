// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch popular photos 1`] = `
Object {
  "popular": Array [],
  "user": Object {
    "description": Array [],
    "name": "",
    "photos": Array [],
    "thumbnail": Object {},
  },
}
`;

exports[`fetch user 1`] = `
Object {
  "popular": Array [],
  "user": Object {
    "description": Array [],
    "name": "",
    "photos": Array [],
    "thumbnail": Object {},
  },
}
`;
